### PR TITLE
POC: wallet connect [WIP]

### PR DESCRIPTION
**Pull request by aleacevedo**
# POC: wallet connect [WIP]

## Summarize

Proof of Concept of wallet connect implementation in mobile

### CODE REVIEW

File: src/helpers/navigation.js [hunk 1]
```diff
@@ -1,18 +1,145 @@
...
+  } else {
+    action();
   }
 }
 
-export function push(...args) {
-  if (navigationRef.isReady()) {
-    navigationRef.dispatch(StackActions.push(...args));
+export function onWillPop() {
+  poppingCounter.isClosing = true;
```
[the remark this answers is not in the document]
Comment (aleacevedo) [reply src/helpers/navigation.js:28]: But it is an object, you can modify an const object without problem

File: src/helpers/navigation.js [hunk 2]
```diff
@@ -1,18 +1,145 @@
...
+}
+
+export function onDidPop() {
+  poppingCounter.isClosing = false;
```
[the remark this answers is not in the document]
Comment (aleacevedo) [reply src/helpers/navigation.js:32]: Same

File: src/helpers/navigation.js [hunk 3]
```diff
@@ -1,18 +1,145 @@
...
+  createNavigationContainerRef,
 } from '@react-navigation/native';
+import { get } from 'lodash';
+import React from 'react';
+import { Value } from 'react-native-reanimated';
+import { useCallbackOne } from 'use-memo-one';
+import { NATIVE_ROUTES } from '../navigation/Routes';
 
-export const navigationRef = createNavigationContainerRef();
+export let TopLevelNavigationRef = createNavigationContainerRef();
```
[the remark this answers is not in the document]
Comment (aleacevedo) [reply src/helpers/navigation.js:14]: I change the value at line 135

File: src/redux/store.js [hunk 4]
```diff
@@ -88,3 +100,28 @@ export const keyringStorage = {
...
+    AsyncStorage.getItem(WALLETCONNECTKEY)
+      .then(savedValues => (savedValues ? JSON.parse(savedValues) : {}))
+      .then(parsedValues =>
+        AsyncStorage.setItem(
+          WALLETCONNECTKEY,
+          Flatted.stringify({ ...parsedValues, values }),
+        ),
+      ),
+  clear: AsyncStorage.setItem(WALLETCONNECTKEY, Flatted.stringify({})),
+};
```
[the remark this answers is not in the document]
Comment (aleacevedo) [reply src/redux/store.js:127]: I think so

File: src/constants/walletconnect.js [hunk 5]
```diff
@@ -0,0 +1,6 @@
+export const PLUG_DESCRIPTION = {
+  description: 'Plug IC Wallet ',
+  icons: [],
```
Comment (mattgle) [inline src/constants/walletconnect.js:3]: Add comment to add plug's icon url

File: src/helpers/walletConnect.js [hunk 6]
```diff
@@ -0,0 +1,48 @@
+import { Buffer } from 'buffer';
+import { blobFromUint8Array, blobToUint8Array } from '@dfinity/candid';
+
+import { getAssets, sign } from '../redux/slices/user';
```
Comment (mattgle) [inline src/helpers/walletConnect.js:4]: use alias

File: src/helpers/walletConnect.js [hunk 7]
```diff
@@ -0,0 +1,48 @@
+import { Buffer } from 'buffer';
```
Comment (mattgle) [inline src/helpers/walletConnect.js:1]: Now we use utils insstead of helpers, move this to utils

File: src/helpers/walletConnect.js [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+    const result = await dispatch(
+      sign({ msg: blobFromUint8Array(new Uint8Array(argsDecoded)) }),
+    ).unwrap();
+    return { result: result.response.toString('base64') };
+  },
+  getAssets: async (args, dispatch) => {
+    const result = await dispatch(getAssets(args));
+    return result;
+  },
+};
```
Comment (mattgle) [inline src/helpers/walletConnect.js:42]: Are those comments ok?

File: src/navigation/index.js [hunk 9]
```diff
@@ -17,20 +17,26 @@ import { navigationRef } from '@/utils/navigation';
...
-  const dispatch = useDispatch();
+const Navigator = ({ routingInstrumentation }, navigationRef) => {
   const { isInitialized, isUnlocked } = useSelector(state => state.keyring);
+  const dispatch = useDispatch();
   let timeoutId = null;
 
   const handleLockState = () => {
     dispatch(setUnlocked(false));
     timeoutId = null;
+    navigationRef?.navigate(Routes.LOGIN_SCREEN);
```
Comment (mattgle) [inline src/navigation/index.js:34]: delete this

File: src/navigation/navigators/SwipeNavigator.js [hunk 10]
```diff
@@ -8,13 +8,21 @@ import ProfileScreen from '@/screens/tabs/Profile';
 import Tokens from '@/screens/tabs/Tokens';
 
 import Routes from '../Routes';
+import useDeepLink from '../../hooks/useDeepLink';
```
Comment (mattgle) [inline src/navigation/navigators/SwipeNavigator.js:11]: use alias

File: src/screens/WalletConnect/callRequest.js [hunk 11]
```diff
@@ -0,0 +1,166 @@
+import React, { useCallback, useState } from 'react';
```
[the remark this answers is not in the document]
Comment (mattgle) [reply src/screens/WalletConnect/callRequest.js:1]: Check other components or screens to see how we handle this

File: src/screens/WalletConnect/index.js [hunk 12]
```diff
@@ -0,0 +1,74 @@
...
+    <Container>
+      <View style={styles.container}>
+        {!Object.keys(meta).length ? (
+          <ActivityIndicator size="large" />
+        ) : (
+          <>
+            <Image source={Plug} style={styles.plugIcon} />
+            <Text style={styles.title}>Wallet Connect</Text>
+            <Text style={styles.title}>{`DAP URL: ${dappUrl}`}</Text>
+            <Text style={styles.title}>{`DAP NAME: ${dappName}`}</Text>
```
Comment (mattgle) [inline src/screens/WalletConnect/index.js:61]: this can be improved

File: src/screens/WalletConnect/styles.js [hunk 13]
```diff
@@ -0,0 +1,47 @@
+import { StyleSheet } from 'react-native';
+
+import { Colors } from '../../constants/theme';
```
Comment (mattgle) [inline src/screens/WalletConnect/styles.js:3]: use aliases

File: src/screens/WalletConnect/styles.js [hunk 14]
```diff
@@ -0,0 +1,47 @@
...
+    alignItems: 'center',
+    justifyContent: 'center',
+    textAlign: 'center',
+    padding: 30,
+    width: '100%',
+    flex: 1,
+  },
+  plugIcon: {
+    height: 116,
+    resizeMode: 'contain',
```
Comment (mattgle) [inline src/screens/WalletConnect/styles.js:16]: check if this looks well on iphone SE and iphone X

File: src/utils/deepLink.js [hunk 15]
```diff
@@ -0,0 +1,48 @@
+import qs from 'qs';
+import URL from 'url-parse';
+
+import { store } from '../redux/store';
+import {
+  walletConnectSetPendingRedirect,
+  walletConnectOnSessionRequest,
+  walletConnectRemovePendingRedirect,
+} from '../redux/slices/walletconnect';
```
Comment (mattgle) [inline src/utils/deepLink.js:9]: use aliases

File: src/utils/deepLink.js [hunk 16]
```diff
@@ -0,0 +1,48 @@
...
+          const type = status === 'approved' ? 'connect' : status;
+          dispatch(
+            walletConnectRemovePendingRedirect({ type, scheme: dappScheme }),
+          );
+        },
+      }),
+    );
+  } else {
+    // This is when we get focused by WC due to a signing request
+  }
```
Comment (mattgle) [inline src/utils/deepLink.js:29]: Can we do something here? Or maybe add a more detail comment

File: src/utils/deepLink.js [hunk 17]
```diff
@@ -0,0 +1,48 @@
...
+  const urlObj = new URL(url);
+  if (urlObj.protocol === 'https:') {
+    const action = urlObj.pathname.split('/')[1];
+    switch (action) {
+      case 'wc': {
+        const { uri } = qs.parse(urlObj.query.substring(1));
+        handleWalletConnect(uri);
+        break;
+      }
+    }
```
Comment (mattgle) [inline src/utils/deepLink.js:46]: is this ok????????

**Review by mattgle — CHANGES_REQUESTED** [review]
Great work!  A few comments!

**Pull request merged**